langchain/src/chat_models/configurable.ts

```diff
@@ -0,0 +1,877 @@
+import { ChatResult } from "@langchain/core/outputs";
+
+// TODO: remove once `EventStreamCallbackHandlerInput` is exposed in core.
+interface EventStreamCallbackHandlerInput
+  extends Omit<LogStreamCallbackHandlerInput, "_schemaFormat"> {}
+
+const _SUPPORTED_PROVIDERS = [
+  "openai",
+  "anthropic",
+  "azure_openai",
+  "cohere",
+  "google_vertexai",
+  "google_genai",
+  "ollama",
+  "together",
+  "fireworks",
+  "mistralai",
+  "groq",
+  "bedrock",
+] as const;
+
+export type ChatModelProvider = (typeof _SUPPORTED_PROVIDERS)[number];
+
+// This is needed so when `bindTools` is called on a configurable model, the typing is correct.
+abstract class BaseChatModelWithBindTools<
+  CallOptions extends BaseChatModelCallOptions = BaseChatModelCallOptions,
+  OutputMessageType extends AIMessageChunk = AIMessageChunk
+> extends BaseChatModel<CallOptions, OutputMessageType> {
+  abstract override bindTools(
+    _tools: (
+      | StructuredToolInterface
+      | Record<string, unknown>
+      | ToolDefinition
+      | RunnableToolLike
+    )[],
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    _kwargs?: Record<string, any>
+  ): Runnable<
+    BaseLanguageModelInput,
+    OutputMessageType,
+    this["ParsedCallOptions"]
+  >;
+
+  _llmType(): string {
+    return "chat_model";
+  }
+
+  abstract _generate(
+    _messages: BaseMessage[],
+    _config?: this["ParsedCallOptions"],
+    _runManager?: CallbackManagerForLLMRun
+  ): Promise<ChatResult>;
+}
+
+export interface ConfigurableChatModelCallOptions
+  extends BaseChatModelCallOptions {
+  tools?: (
+    | StructuredToolInterface
+    | Record<string, unknown>
+    | ToolDefinition
+    | RunnableToolLike
+  )[];
+}
+
+async function _initChatModelHelper(
+  model: string,
+  modelProvider?: string,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  params: Record<string, any> = {}
+): Promise<BaseChatModelWithBindTools> {
+  const modelProviderCopy = modelProvider || _attemptInferModelProvider(model);
+  if (!modelProviderCopy) {
+    throw new Error(
+      `Unable to infer model provider for { model: ${model} }, please specify modelProvider directly.`
+    );
+  }
+
+  try {
+    switch (modelProviderCopy) {
+      case "openai": {
+        const { ChatOpenAI } = await import("@langchain/openai");
+        return new ChatOpenAI({ model, ...params });
+      }
+      case "anthropic": {
+        const { ChatAnthropic } = await import("@langchain/anthropic");
+        return new ChatAnthropic({ model, ...params });
+      }
+      case "azure_openai": {
+        const { AzureChatOpenAI } = await import("@langchain/openai");
+        return new AzureChatOpenAI({ model, ...params });
+      }
+      case "cohere": {
+        const { ChatCohere } = await import("@langchain/cohere");
+        return new ChatCohere({ model, ...params });
+      }
+      case "google_vertexai": {
```

Comment:
Is this for Python parity? Would double check

Otherwise would make slightly more sense to do `google-vertexai` to match the package names

Reply:
yes py uses `google_vertexai`. I'll update to allow both

Reply:
eh actually, I'm just going to make it only `google-vertexai`/`google-genai`. Exact parity for that field isn't very important.